core/src/main/scala/com.snowplowanalytics.stream/loader/sinks/KinesisSink.scala

```diff
@@ -127,3 +134,36 @@ class KinesisSink(conf: KinesisSinkConfig) extends ISink {
       if (cond) f(builder) else builder
   }
 }
+
+object KinesisSink {
+
+  // Represents a partition key and the serialized record content
+  private type KeyedData = (String, Array[Byte])
+
+  /**
+   *  Takes a list of records and splits it into several lists, where each list is as big as
+   *  possible with respecting the record limit and the size limit.
+   */
+  def groupOutputs(conf: KinesisSinkConfig)(keyedData: List[KeyedData]): List[List[KeyedData]] = {
```

Comment:
The function could take just `recordLimit` and `sizeLimit` as parameters, so that you don't have to deal with `KinesisSinkConfig` in the test

Reply:
I considered that!  (I know it's how you did it in enrich).  I will explain my reasoning for doing it differently:

If `groupOutputs` has this signature:

```
def groupOutputs(recordLimit: Int, sizeLimit: Int, keyedData: List[KeyedData]): List[List[KeyedData]]
```

then it's all too easy to make this mistake when calling the function:

```
groupOutputs(config.sizeLimit, config.recordLimit, keyedData)
```

The compiler doesn't tell me that I got my first two parameters the wrong way around!

It's a close call, but in the end I decided I preferred having the safer method signature even if it made the test more clumsy.

Reply:
Just for fun... an alternative way could by having the config implement a trait `BatchLimits`

```
trait BatchLimits {
  def: sizeLimit: Int
  def: recordLimit: Int
}

case class KinesisSinkConfig(
  sizeLimit: Int,
  recordLimit: Int,
  // Lots of other fields go here
) extends BatchLimits

def groupOutputs(batchLimits: BatchLimits, keyedData: List[KeyedData]): List[List[KeyedData]] = ???

groupOutputs(config, keyedData)
```

And in the tests:

```
val limits = new Batch Limits {
  val sizeLimit = 10
  val recordLimit = 10
}

val result = groupOutputs(limits, inputs)
```

Reply:
> The compiler doesn't tell me that I got my first two parameters the wrong way around!

It doesn't tell you either that you're right or wrong when you use `conf.recordLimit` and `conf.sizeLimit` inside the `foldLeft`, you could just as well switch them.

Reply:
I would prefer the alternative with `BatchLimits` to passing the whole `KinesisConfig` to `group`

Reply:
> It doesn't tell you either that you're right or wrong when you use conf.recordLimit and conf.sizeLimit inside the foldLeft, you could just as well switch them.

True!  But... but I find this is more of an obvious mistake.  Because just by looking at the implementation of `groupOutputs` I can see if I've made a mistake.  Whereas the other way around, I need to look in two different places simultaneously to see if I've made a mistake (the call site and the method signature).

It's a very weak argument though.  I only have a slight preference for doing it my suggested way.

> I would prefer the alternative with BatchLimits to passing the whole KinesisConfig to group

Let's do this!  I think it will be quite neat.